src/v/cluster/rm_stm.cc

```diff
@@ -2611,6 +2613,45 @@ ss::future<> rm_stm::handle_eviction() {
       });
 }
 
+void rm_stm::spawn_background_clean_for_pids(rm_stm::clear_type type) {
+    switch (type) {
+    case rm_stm::clear_type::tx_pids: {
+        if (
+          _log_state.fence_pid_epoch.size() <= _max_concurrent_producer_ids()) {
+            return;
+        }
+        break;
+    }
+    case rm_stm::clear_type::idempotent_pids: {
+        if (
+          _log_state.idempotent_pids_order.size()
+          <= _max_concurrent_producer_ids()) {
+            return;
+        }
+        break;
+    }
+    }
+
+    ssx::spawn_with_gate(_gate, [this, type] {
+        return _state_lock.hold_read_lock().then(
+          [this,
+           type](ss::basic_rwlock<>::holder unit) mutable -> ss::future<> {
+              switch (type) {
+              case rm_stm::clear_type::tx_pids: {
+                  vlog(_ctx_log.trace, "Gonna clear old tx pids");
+                  co_await clear_old_tx_pids();
+                  break;
+              }
+              case rm_stm::clear_type::idempotent_pids: {
+                  vlog(_ctx_log.trace, "Gonna clear old idempotent pids");
+                  co_await clear_old_idempotent_pids();
+                  break;
+              }
+              }
```

Comment:
Avoid capturing lambdas that are coroutines

Reply:
Why? 
I think I am not understand idea of it

Reply:
in this case it is safe but basically they are unfsafe as all the refrences may be invalidated after first co_await in lambda